plugins/search-react/src/extensions.tsx

```diff
@@ -0,0 +1,228 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {
+  ReactNode,
+  createElement,
+  isValidElement,
+  useCallback,
+} from 'react';
+
+import {
+  getComponentData,
+  useElementFilter,
+  createReactExtension,
+  useAnalytics,
+} from '@backstage/core-plugin-api';
+import {
+  Result,
+  SearchResult,
+  SearchDocument,
+} from '@backstage/plugin-search-common';
+
+import { DefaultResultListItem } from './components';
+import { List, ListProps } from '@material-ui/core';
+
+/**
+ * @internal
+ * Key for result extensions.
+ */
+const SEARCH_RESULT_LIST_ITEM_EXTENSION =
+  'search.results.list.items.extensions.v1';
+
+/**
+ * @internal
+ * Returns the first extension element found for a given result, and null otherwise.
+ * @param elements - All extension elements.
+ * @param result - The search result.
+ */
+const findSearchResultListItemExtensionElement = (
+  elements: ReactNode[],
+  result: SearchResult,
+) => {
+  for (const element of elements) {
+    if (!isValidElement(element)) continue;
+    const data = getComponentData<
+      Pick<SearchResultListItemExtensionOptions, 'predicate' | 'component'>
+    >(element, SEARCH_RESULT_LIST_ITEM_EXTENSION);
+    if (!data) continue;
+    const { predicate, component } = data;
+    if (!predicate?.(result)) continue;
+    return createElement(component, {
+      rank: result.rank,
+      highlight: result.highlight,
+      result: result.document,
+      noTrack: true,
+      // Use props over the context in situations where a consumer is manually rendering the extension
+      ...element.props,
+    });
+  }
+  return null;
+};
+
+/**
+ * @public
+ * Props for extension options component.
+ */
+export type SearchResultListItemExtensionOptionsComponentProps<
+  P extends {} = {},
+  D extends SearchDocument = SearchDocument,
+> = P & {
+  rank?: Result<D>['rank'];
+  highlight?: Result<D>['highlight'];
+  result: Result<D>['document'];
+};
+
+/**
+ * @public
+ * Make default extension options component props optional.
+ */
+export type SearchResultListItemExtensionComponent<
+  P extends SearchResultListItemExtensionOptionsComponentProps,
+> = (
+  props: Omit<P, keyof SearchResultListItemExtensionOptionsComponentProps> &
+    Partial<SearchResultListItemExtensionOptionsComponentProps>,
+) => JSX.Element | null;
+
+/**
+ * @public
+ * Options for {@link createSearchResultListItemExtension}.
+ */
+export type SearchResultListItemExtensionOptions<
+  T extends SearchResultListItemExtensionOptionsComponentProps = SearchResultListItemExtensionOptionsComponentProps,
+> = {
+  /**
+   * The extension name.
+   */
+  name: string;
+  /**
+   * The extension component.
+   */
+  component: (props: T) => JSX.Element | null;
```

Comment:
How come you didn't end up supporting lazy loaded components here? Are you expecting these components to always be very small?

Reply:
We're passing a component that always returns null to createReactExtension [here](https://github.com/backstage/backstage/pull/15855/files#diff-4c59e19e90caafd5ccb90cb3236950aa33e04d44d061b4b98aa4cbd3bda8f295R132), would you recommend using `lazy` for this case?

Reply:
No but you could make the the component itself that is registered in the component data be lazy loaded. Not saying that it's necessarily better, scaffolder field extensions are sync for example. Just wanna make sure it's considered. I don't know myself if search items are large enough for it to make sense. It can also be added later on by adding more options, so it's mostly for the initial implementation and the existing extensions.

Reply:
The components are list items and can be as simple as the default component that we export as having a preview and querying external sources as our internal use case, so I changed it to lazy components; thanks for bringing that up.